### PR TITLE
Initial FTS integration code for walrep.

### PR DESCRIPTION
**Pull request by ashwinstar**
The main goal of this PR is to detect and reflect mirror's status up (or down) via primary, on master in catalog.

To achieve the above goal:

- Adding IsMirrorUp() function which walks through WalSndCtl and based on
  walsnds state, reflects mirror up or down. So, now this is the only code
  touching the mirror status in gp_segment_configuration. Both up and mirror
  will get detected via FTS hence utility just needs to start the mirror and FTS
  would perform the catalog updates.

- Currently added FIXME as cannot acquire LWLock in IsMirrorUp(). This
  restriction will be removed once ftsprobe handler is coded as a lightweigth
  backend process like walsender, instead of processing FTS probes on primary in
  ProcessStartupPacket() directly.

- Adding separate file for FTS probe handler code for walrep instead of
  currently present in postmaster.c

- Basic framework for master using the fts process to probe each primary and
  gets the status back. New response structures are defined to ease future
  development for adding support for detecting sync, not-in-sync.

- Persist the mirror status change to catalog tables. This also does the status
  of the primary. Coded to prevent status change to mirror if the primary is
  offline.

- Adding function to request / trigger FTS probe.

Reference to related mailing list discussion:
https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/sRAjNHHTqMA/Uzi-0GR4AQAJ

From,
Shin, Taylor, Ashwin

### CODE REVIEW

File: src/include/postmaster/fts.h [hunk 1]
```diff
@@ -90,6 +113,10 @@ typedef struct
 	uint32 stateMirror;
 } FtsSegmentPairState;
 
+typedef struct ProbeResponse
```
Comment (asimrp) [inline src/include/postmaster/fts.h]: Missing `#ifdef USE_SEGWALREP`?

File: src/backend/fts/ftsprobe.c [hunk 2]
```diff
@@ -490,25 +609,31 @@ static bool
...
 				  probeInfo->dbId, probeInfo->conn->errorMessage.data);
 		return false;
 	}
-	/* segment responded to probe, mark it as alive */
-	probeInfo->segmentStatus = PROBE_ALIVE;
-	write_log("FTS: segment (content=%d, dbid=%d) reported segmentstatus %x to the prober.",
-			  probeInfo->segmentId, probeInfo->dbId, probeInfo->segmentStatus);
 #ifdef USE_SEGWALREP
 	/*
 	 * Walrep primary segments return empty response to FTS probes, which means
```
Comment (asimrp) [inline src/backend/fts/ftsprobe.c]: This comment is now obsolete.

File: src/backend/fts/ftsprobe.c [hunk 3]
```diff
@@ -301,9 +288,141 @@ probeSegment(CdbComponentDatabaseInfo *dbInfo)
...
+			break;
+
+		/*
+		 * If FTS is shutdown, abort.
+		 */
+		if (!FtsIsActive())
+			break;
+
+		/* now let's probe the primary. */
+		probe_response_per_segment *response = context->responses + response_index;
```
Comment (asimrp) [inline src/backend/fts/ftsprobe.c]: Let's use `&contex->responses[response_index]` for consistency.

File: src/backend/fts/ftsprobe.c [hunk 4]
```diff
@@ -234,24 +235,8 @@ FtsProbeSegments(CdbComponentDatabases *dbs, uint8 *probeRes)
 	}
 }
 
-/*
- * This is called from several different threads: ONLY USE THREADSAFE FUNCTIONS INSIDE.
- */
-static char
-probeSegment(CdbComponentDatabaseInfo *dbInfo)
+static char probeSegmentHelper(CdbComponentDatabaseInfo *dbInfo, ProbeConnectionInfo probeInfo)
```
Comment (asimrp) [inline src/backend/fts/ftsprobe.c]: Let's stick to PostgreSQL convention of return type on its own line.

File: src/backend/fts/ftsprobe.c [hunk 5]
```diff
@@ -301,9 +288,141 @@ probeSegment(CdbComponentDatabaseInfo *dbInfo)
 
 	if (probeInfo.conn)
 		PQfinish(probeInfo.conn);
+
+#ifndef USE_SEGWALREP
```
Comment (asimrp) [inline src/backend/fts/ftsprobe.c]: In `USE_SEGWALREP` case, wouldn't this cause a compiler warning, that the function is defined to return char but returns void?

Comment (ashwinstar) [reply src/backend/fts/ftsprobe.c]: If you look at full set of changes, this is fixed with code movement commit.

**Pull request merged**